common/src/main/java/com/wynntils/core/text/StyledTextPart.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.text;
+
+import com.wynntils.utils.wynn.WynnUtils;
+import java.util.Objects;
+import java.util.function.Function;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.Style;
+
+public final class StyledTextPart {
+    private final String text;
+    private PartStyle style;
+
+    private final StyledText parent;
+
+    StyledTextPart(String text, Style style, StyledText parent, Style parentStyle) {
+        this.parent = parent;
+        this.text = text;
+
+        // Must be done last
+        this.style = PartStyle.fromStyle(style, this, parentStyle);
+    }
+
+    StyledTextPart(StyledTextPart part, StyledText parent) {
+        this.text = part.text;
+        this.style = part.style;
+        this.parent = parent;
+    }
+
+    // This factory is used to create a StyledTextPart from a string that has formatting codes
+    // It is separate from the constructor because this only needs to be applied in cases there the text could have
+    // formatting codes
+    static StyledTextPart fromStyledString(String text, Style style, StyledText parent, Style parentStyle) {
```

Comment:
So maybe I'm missing something here since I have not looked at StyledText itself yet, but this sounds like it is on the wrong level. If you are creating a part of a string like "§cfoo", then I guess this works, but then you only need to check if the first character is §. But if you have a string like "§afoo§cbar" or even "foo§cbar", then you would in fact need to create two different StyledTextParts for that single component literal.

And it looks like this function sort-of tries to do that, but not correctly. I think it is the responsibility of StyledText to make sure it only tries to instantiate StyledTextParts that has the very basic minimal "atom" of a stretch of text -- without any formatting codes!  -- and a style that applies to the entire part.

Reply:
If this comment seem totally wrong to you, ignore it and I'll continue reading the code tomorrow. But if it makes sense, feel free to go ahead and fix.

Reply:
Well, this is the "handle if the component has formatting" code. You are correct in the regard that once we not find a format, we can "break" out of the loop.

This is NOT the code which is called from `StyledText#fromString`. There, it is handled correctly, and this is skipped, since there is no case where we have styles in any part.

But when parsing from components, or after doing string operations, like append, we need to check if we added new styles as string.

Reply:
I pushed the bit more optimized, and more understandable code. I also renamed it `fromComponentWithPossibleFormattingCodes` so it is not confused anymore.

Reply:
I'll resolve this thread since we are discussion the same problem elsewhere.